pydatastructs/trees/heaps.py

```diff
@@ -84,9 +84,15 @@ def __new__(cls, elements=None, heap_property="min", d=4):
             raise ValueError("%s is invalid heap property"%(heap_property))
         if elements is None:
             elements = DynamicOneDimensionalArray(TreeNode, 0)
-        else:
+        elif type(elements) in (list,tuple):
             if not all(map(lambda x: _check_type(x, TreeNode), elements)):
-                raise ValueError("Expect a list/tuple of TreeNode got %s"%(elements))
+                raise ValueError("Expected a list/tuple/Array of TreeNode got %s"%(elements))
+            elements = DynamicOneDimensionalArray(TreeNode, len(elements), elements)
+        elif _check_type(elements, Array):
+            if(not _check_type(elements[0], TreeNode)):
+                raise ValueError(f'Expected a list/tuple/Array of TreeNode got {type(elements[0])}')
+        else:
+            raise ValueError(f'Expected a list/tuple/Array of TreeNode got {type(elements)}')

```

Comment:
Remove the type checks here. If some one keeping something of not of type `TreeNode`, it will be checked while filling `DODA`(type checks are there in it). 
Rest looks okay, I misinterpreted the code in `master`.

Reply:
If elements are not list/tuple then they can simply be converted to DODA, after that part code can safely assume that elements is DODA object and work on the basis of that. I hope you are getting my point.

Reply:
so just put it in DODA and let it give errors if any?

Reply:
It is first getting compared by `obj._comp` and then sending it to `DODA` so if I remove the type checks it doesn't pass the tests

Reply:
```
{
	"resource": "/D:/projects/pydatastructs/pydatastructs/trees/tests/test_heaps.py",
	"owner": "pytest",
	"code": "./pydatastructs/trees/tests/test_heaps.py::test_BinaryHeap",
	"severity": 8,
	"message": "Fail: TypeError: '<=' not supported between instances of 'tuple' and 'int'",
	"source": "pytest",
	"startLineNumber": 8,
	"startColumn": 5,
	"endLineNumber": 8,
	"endColumn": 20,
	"relatedInformation": [
		{
			"startLineNumber": 8,
			"startColumn": 5,
			"endLineNumber": 8,
			"endColumn": 20,
			"message": "test_BinaryHeap",
			"resource": "/d:/projects/pydatastructs/pydatastructs/trees/tests/test_heaps.py"
		},
		{
			"startLineNumber": 56,
			"startColumn": 5,
			"endLineNumber": 56,
			"endColumn": 38,
			"message": "assert raises(ValueError, lambda:",
			"resource": "/d:/projects/pydatastructs/pydatastructs/trees/tests/test_heaps.py"
		},
		{
			"startLineNumber": 16,
			"startColumn": 9,
			"endLineNumber": 16,
			"endColumn": 15,
			"message": "code()",
			"resource": "/d:/projects/pydatastructs/pydatastructs/utils/raises_util.py"
		},
		{
			"startLineNumber": 56,
			"startColumn": 5,
			"endLineNumber": 56,
			"endColumn": 38,
			"message": "assert raises(ValueError, lambda:",
			"resource": "/d:/projects/pydatastructs/pydatastructs/trees/tests/test_heaps.py"
		},
		{
			"startLineNumber": 262,
			"startColumn": 9,
			"endLineNumber": 262,
			"endColumn": 61,
			"message": "obj = DHeap.__new__(cls, elements, heap_property, 2)",
			"resource": "/d:/projects/pydatastructs/pydatastructs/trees/heaps.py"
		},
		{
			"startLineNumber": 96,
			"startColumn": 9,
			"endLineNumber": 96,
			"endColumn": 21,
			"message": "obj._build()",
			"resource": "/d:/projects/pydatastructs/pydatastructs/trees/heaps.py"
		},
		{
			"startLineNumber": 104,
			"startColumn": 13,
			"endLineNumber": 104,
			"endColumn": 29,
			"message": "self._heapify(i)",
			"resource": "/d:/projects/pydatastructs/pydatastructs/trees/heaps.py"
		},
		{
			"startLineNumber": 122,
			"startColumn": 21,
			"endLineNumber": 122,
			"endColumn": 88,
			"message": "target = j if self._comp(self.heap[j].key, self.heap[target].key) \\",
			"resource": "/d:/projects/pydatastructs/pydatastructs/trees/heaps.py"
		},
		{
			"startLineNumber": 80,
			"startColumn": 13,
			"endLineNumber": 80,
			"endColumn": 78,
			"message": "obj._comp = lambda key_parent, key_child: key_parent <= key_child",
			"resource": "/d:/projects/pydatastructs/pydatastructs/trees/heaps.py"
		}
	]
}
```
I removed this line expecting it to be checked in DODA
```
elif type(elements) in (list,tuple):
            """
            if not all(map(lambda x: _check_type(x, TreeNode), elements)):
                raise ValueError("Expected a list/tuple/Array of TreeNode got %s"%(elements))
            """
            elements = DynamicOneDimensionalArray(TreeNode, len(elements), elements)
```
but it fails

Reply:
Try applying [this diff file](https://github.com/codezonediitj/pydatastructs/files/4408877/pds_diff.txt).
Do, `git apply pds_diff.txt`. Do it only if you haven't applied any commits on top of your last commits.

Reply:
ok

Reply:
patch failed

Reply:
shall I try it on master?

Reply:
I will push the changes to your branch. Just enable `Edits from maintainer`.

Reply:
![image](https://user-images.githubusercontent.com/42893005/78029495-52c5f500-737e-11ea-91ce-80f8c158d858.png)
already enabled sir